File: 1485-clone-binary-tree-with-random-pointer/1485-clone-binary-tree-with-random-pointer.cpp
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     Node *left;
 *     Node *right;
 *     Node *random;
 *     Node() : val(0), left(nullptr), right(nullptr), random(nullptr) {}
 *     Node(int x) : val(x), left(nullptr), right(nullptr), random(nullptr) {}
 *     Node(int x, Node *left, Node *right, Node *random) : val(x), left(left), right(right), random(random) {}
 * };
 */

class Solution {
public:
    unordered_map<Node*,NodeCopy*> map;
    NodeCopy* copy(Node* node) {
        if(!node) return nullptr;
        
        NodeCopy* root = nullptr;
        if(map.find(node) == map.end()) {
            root = new NodeCopy(node->val);
            map[node] = root;
        }
        else {
            root = map.at(node);
        }
        
        if(node->random){
            if(map.find(node->random) == map.end()){
                root->random = new NodeCopy(node->random->val);
                map[node->random] = root->random;
            }
            else{
                root->random = map.at(node->random);   
            }
        }
        
        root->left = copy(node->left);
        root->right = copy(node->right);
        
        return root;
    }
    
    NodeCopy* copyRandomBinaryTree(Node* root) {
        
        return copy(root);
        
    }
};

File: 1485-clone-binary-tree-with-random-pointer/1485-clone-binary-tree-with-random-pointer.java
/**
 * Definition for Node.
 * public class Node {
 *     int val;
 *     Node left;
 *     Node right;
 *     Node random;
 *     Node() {}
 *     Node(int val) { this.val = val; }
 *     Node(int val, Node left, Node right, Node random) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *         this.random = random;
 *     }
 * }
 */

class Solution {
    HashMap<Node,NodeCopy> map;
    public NodeCopy copy(Node node) {
        if(node==null) return null;
        
        NodeCopy root = map.getOrDefault(node, new NodeCopy(node.val));
        map.put(node,root);
        
        if(node.random!=null) {
            root.random = map.getOrDefault(node.random,new NodeCopy(node.random.val));   
            map.put(node.random,root.random);
        }
        
        root.left = copy(node.left);
        root.right = copy(node.right);
        
        return root;
    }
    
    public NodeCopy copyRandomBinaryTree(Node root) {
        map = new HashMap<>();
        return copy(root);
    }
}